Reduce use of Quadrature objects in Manifolds.

gassmoeller opened the pull request:
This should help with #2896 as discussed in #2990. 
@kronbichler is this approximately what you imagined?
For now I only deprecated the old functions, maybe some of them are not used from the outside, and can be removed immediately? Not sure about that.
I have not checked the speedup for other Manifolds, but for the SphericalManifold not using the `Quadrature` in `get_new_point` reduced its compute cycles by about 20 %, and there are more savings further up in the calling functions that do not longer construct Quadrature objects (get_intermediate_points in total saves 30% of its cycles). For simpler Manifolds this might be even higher, because their `get_intermediate_point` functions are likely cheaper.
Before merging this PR I would like to do some tests in 3D first, to make sure the O(N^2) sort in `get_new_point` does not outscale the benefits of not allocating memory, but this is not relevant for any other Manifold than `SphericalManifold`.

gassmoeller commented:
Ok seems like the results are consistent. 3D sees a bit less of a speedup, but maybe that is because the calculation of the position is more expensive compared to the memory allocation. Anyway, this is ready for a review.

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -79,7 +79,54 @@ namespace Manifolds
+  template <typename MeshIteratorType>
+  std::pair<std::vector<Point<MeshIteratorType::AccessorType::space_dimension> >,
+      std::vector<double> >
+      get_default_points_weights(const MeshIteratorType &iterator,
```
maybe call it `get_default_points_and_weights`?

gassmoeller commented:
I renamed the function and fixed another small bug. This should be good to go from my perspective (after @kronbichler has taken a look to make sure I did not screw up things).

gassmoeller commented:
Ping. And could somebody start the tester?
This is really less code to review than it looks like. All new functions are simply copies of existing ones with `Quadrature` objects replaced by vectors of points and weights. The only algorithmic change is that `Manifold::get_new_point()` not longer generates a `QSorted` object, but instead figures out the order of the points and weights itself, and then does the pair-wise reduction in that order.

bangerth commented:
/run-tests

I would like @kronbichler or @luca-heltai to take a look.

luca-heltai commented:
This looks good to me. I'm a bit skeptical about the O(n^2) in the sorting. From the theoretical point of view, the Sorting is required to ensure stability of the algorithm, but in what I saw so far this is not influential in most cases. Grids are refined in a slightly different way when no sorting is applied, but the quality of the mesh is comparable. 

May be we could simply remove the sorting? Notice that sorting is not necessary, for example, if the Manifold is intrinsically a map (in this case the get_new_point is associative).

@kronbichler , how much was the influence of sorting in your benchmarks? I think that the squeeze is not worth the juice...

kronbichler commented on source/grid/manifold.cc:
```diff
@@ -87,6 +87,68 @@ get_new_point (const Quadrature<spacedim> &quad) const
+  // overhead and use a very simple search of O(N^2) to find the order in which
+  // to loop over points.
+  std::vector<unsigned int> permutation(n_points);
+  std::vector<bool> found(n_points,false);
```
Do we know for sure that the number of points does not exceed 26? Then we should make those two arrays fix-size arrays so they are put on the stack rather than allocated. (Alternatively, we can put them into a `ThreadLocalStorage` object, but that sounds a bit exaggerated for 130 bytes of data.)

gassmoeller replied:
Unfortunately no, we can not be entirely sure. This function is public and in principle users can call the function with any number of points and weights. However, any call from within deal.II hands over at most 26 points (the number of points generated by `get_defaults_points_and_weights` in 3D). I can not think of an application that would want to find a new point within a large number of points, but that  does not mean there is no such application. Alternatively, one could also use std::sort with a comparison function that accesses the weights vector (like in the QSorted constructor), but I do not see how to do that without using a lambda function here. In my 3D application the sorting takes about 10% of the total cycles of this function (5% for allocation, 5% for sorting), while 80% are spent in the `get_intermediate_point` function. Not sure about the best way to proceed here.

bangerth replied:
It can be any number of points if you just use a high enough polynomial degree for the mapping. 26 just happens the maximum you see in your application.

gassmoeller replied:
I am not so sure (polite for: I do not think so :smile:). This quadrature object was never used like an actual quadrature object, it just happened to be used as a container for some points and weights. All places in the library fill this object with either:
- the vertices of some object (line, quad, or hex)
- midpoints of above object's lines and faces
- a combination of the above

That means the maximum number of points is `number_of_vertices + number_of_line_midpoints + number_of_faces = 26` for 3D. Actually most places that use this function (including `MappingQ` and `MappingManifold`) only call it with the vertices of the cell (8 in 3D), only `get_default_points_and_weights` uses the larger number.

It is true that MappingQGeneric calls this function more often when using higher order mappings, but it always hands over the same vertices of the cell as surrounding points, it just computes more new points the higher the polynomial degree.

bangerth replied:
I think this function is called for each edge with the adjacent vertices, for each quad with the adjacent vertices and *all* of the points we had previously computed on the edges, and finally for each hex with all of the vertex, edge, and quad points. Because the number of edge and face points depends on the degree of the mapping, I think we can get arbitrarily many points, if I recall correctly how this all works.

gassmoeller replied:
Hmm, I really do not see where this is happening. I have looked through all calling functions, and can not find the place you are referring to. There are places in `MappingManifold` that compute as many new points as necessary for the mapping degree, but it does so by always using the same cell vertices and changing the weights of them for each new quadrature point. And `MappingQ::get_intermediate_points` calls this function with a vector of points, but this vector can only have 2, 4 or 8 points, otherwise it throws an exception. Then there are a number of calling places in `tria_accessor.cc`, but they all call the `get_new_point_on_...` function variants that use the default points and weights (maximum 26 points). Finally there are a lot of calls from `tria_boundary_lib.cc`. Those are pretty confusing, but all of them end up calling `FlatManifold::get_new_point` instead of `Manifold:get_new_point` so they never see this algorithm.
 Also, as a test I have added output and asserts to make sure the number of points never exceeds 26, but still all mapping and manifold tests succeed. I also explicitly made a test with a higher order mapping, and a SphericalManifold but still no call with more than 26 points.
But what about the following: Take a look at #3066. Would that be ok with you?

bangerth replied:
I must be wrong then. But I guess you are probably right. I thought that `MappingQ` computes the interior points by providing *all* of the points on the boundary, but you are probably right that it just takes the vertices. Thanks for checking!

kronbichler commented on source/grid/manifold.cc:
```diff
@@ -274,7 +336,8 @@ Point<spacedim>
 {
-  return get_new_point (get_default_quadrature(line));
+  const std::pair<std::vector<Point<spacedim> >, std::vector<double> > points_weights(get_default_points_and_weights(line));
+  return get_new_point (points_weights.first,points_weights.second);
```
Here and in the two two functions below, we still allocate memory. At some point, we might get around by re-using some data arrays in thread local storage, but I think for now you already reduce the number of memory allocations from 5 to 2 or so.

gassmoeller replied:
Yes reducing this further requires some changes in the places calling these functions. I tried to keep things simple.

kronbichler replied:
Agreed!

kronbichler commented on source/grid/manifold.cc:
```diff
@@ -87,6 +87,68 @@ get_new_point (const Quadrature<spacedim> &quad) const
+Manifold<dim, spacedim>::
+get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
+               const std::vector<double>           &weights) const
+{
```
Can't you redirect the now-deprecated `get_new_point` with quadrature argument to this function by calling `get_new_point(quad.get_points(), quad.get_weights())`? The same on the `get_new_point` methods of the derived classes below.

gassmoeller replied:
You are right. Have done so.

kronbichler commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3542,8 +3540,7 @@ namespace
 
-                    Quadrature<spacedim> quadrature(surrounding_points, w);
-                    points[c]=manifold.get_new_point(quadrature);
+                    points[c]=manifold.get_new_point(surrounding_points, w);
```
very nice simplification in this function!

kronbichler commented:
> May be we could simply remove the sorting? Notice that sorting is not necessary, for example, if the Manifold is intrinsically a map (in this case the get_new_point is associative).
@kronbichler , how much was the influence of sorting in your benchmarks? I think that the squeeze is not worth the juice...

I can run my test once this gets merged. Since there are at most 26 points, I think that the worst-case 26^2 operations is not too bad. Certainly not much worse than *real* sorting algorithms. But if you say that it would be algorithmically ok to skip the sorting, I would definitely not be opposed to that.

bangerth commented:
I think @gassmoeller found that he could already save 90% of the CPU time by just removing the memory allocation. It may be that there is not all that much to save any more by skipping the sorting. I would imagine that most of the time the array to sort is not very large and that one could save only a few instructions.

But I'll let him give us the numbers rather than speculate.

kronbichler commented:
@bangerth I agree that the bad stuff is memory allocation and reducing it is the most important aspect. I measured at some point and memory allocation takes around 150-200 clock cycles of a comtemporary CPU. If you assume vectorized code consisting of fused-multiply-add instructions, you could do 3200 FLOPs in the same time of one memory allocation.

luca-heltai commented:
Ok, then we should avoid all get_new_point_on_XXXX functions, and only call get_new_point(w,q) with w and q allocated in the outer most loop only once.

luca-heltai commented:
I'd start by deprecating the get_new_point_on_XXX (reminiscent of the old Boundary interface), and replacing all their use internally with get_new_point

luca-heltai commented:
Maybe in a followup PR?

kronbichler commented:
@luca-heltai I agree.

tamiko commented:
@gassmoeller Can you please rebase on current master for the merge?

gassmoeller commented:
I rebased and addressed the uncontroversial comments. I guess we can proceed with this PR, and decide about the possible improvements after @kronbichler tested the new state in his application.

kronbichler commented:
I guess we can leave the '26' points to another PR. This looks really good now.

kronbichler commented:
@gassmoeller It seems that this commit makes the test `grid/torus_01` fail because some points are slightly different. Can you have a short look at the changes and whether the new data is OK (it seems so at first sight)? I do not see how the data can change given that you simply did optimizations, though.

cpraveen commented:
I am facing some problems with manifolds/mapping/quadrature points (not sure where exactly) in the recent git versions which could be related to this. I have raised one issue #3013 and another one is described here https://gist.github.com/cpraveen/20c007f7097a5933cfe22fb3744615d9

luca-heltai commented:
@cpraveen , @nicola-giuliani and I have worked on your public gist, to reduce to a minimum. Here is the result:

https://gist.github.com/nicola-giuliani/dc5ce21956b8df4fd8d8d06f7213cfe4

Where we do get to 1e-13 when computing the area of the sphere in a way similar to what you did. These are the errors we get:

~~~
Testing the surface computation of a sphere
exact area : 12.5664
1 8 0.36338 26 0.121441 98 0.0326377 386 0.00830961 1538 0.00208692 6146 0.000522328 
2 26 0.0170433 98 0.00128528 386 7.7814e-05 1538 4.81402e-06 6146 3.00044e-07 24578 1.87392e-08 
3 56 0.00628029 218 8.19733e-05 866 1.27395e-06 3458 1.98772e-08 13826 3.10475e-10 55298 5.05044e-12 
4 98 0.000536104 386 2.05926e-06 1538 7.0732e-09 6146 2.65364e-11 24578 5.10302e-14 98306 3.15935e-13
~~~

The gist you had has too many things in there to really understand what's failing and what's not. We're now trying to extend this to your case, until we get to an error, but so far, things seem working:

https://gist.github.com/nicola-giuliani/86f535ddfd83485150767d123c95c242

results:
~~~
1 386 0.00830961 386 0.00830961 386 0.00830961 386 0.00830961 386 0.00830961 386 0.00830961 386 0.00830961 386 0.00830961 386 0.00830961 386 0.00830961 
2 1538 4.81402e-06 1538 4.81402e-06 1538 4.81402e-06 1538 4.81402e-06 1538 4.81402e-06 1538 4.81402e-06 1538 4.81402e-06 1538 4.81402e-06 1538 4.81402e-06 1538 4.81402e-06 
3 3458 1.98771e-08 3458 1.98772e-08 3458 1.98772e-08 3458 1.98771e-08 3458 1.98772e-08 3458 1.9877e-08 3458 1.98772e-08 3458 1.9877e-08 3458 1.98772e-08 3458 1.98771e-08 
4 6146 2.66224e-11 6146 2.66006e-11 6146 2.66143e-11 6146 2.65629e-11 6146 2.64937e-11 6146 2.647e-11 6146 2.65364e-11 6146 2.66952e-11 6146 2.65688e-11 6146 2.64416e-11 
5 9602 1.52384e-13 9602 7.35062e-15 9602 4.28315e-14 9602 1.9366e-14 9602 5.76741e-14 9602 1.45599e-14 9602 3.95802e-14 9602 9.88092e-14 9602 1.38531e-14 9602 1.09977e-13 
6 13826 6.95481e-14 13826 2.50204e-14 13826 1.00364e-14 13826 4.69309e-14 13826 1.27222e-15 13826 5.54123e-14 13826 1.08846e-14 13826 4.08525e-14 13826 8.62284e-15 13826 3.22296e-14 
7 18818 2.38612e-13 18818 1.24395e-14 18818 2.53031e-14 18818 4.73549e-14 18818 9.64061e-14 18818 1.01778e-14 18818 5.06062e-14 18818 6.72864e-14 18818 4.83444e-14 18818 1.21568e-13 
8 24578 2.38895e-14 24578 1.44185e-14 24578 8.63697e-14 24578 2.21932e-14 24578 1.13228e-13 24578 9.47099e-15 24578 1.88006e-14 24578 4.55173e-14 24578 1.12945e-13 24578 1.08563e-13 
9 31106 3.1749e-13 31106 2.75648e-14 31106 1.15914e-14 31106 4.80617e-14 31106 1.83765e-14 31106 4.93339e-14 31106 1.88006e-14 31106 6.92654e-14 31106 1.31463e-14 31106 5.3716e-15
~~~

We'll try to understand better if the problem is what you pointed out in #3013, but these two gists seem to give correct results.

gassmoeller commented:
@kronbichler: I see the change (some coordinates that should probably be close to 0 are 1e-17 in the reference output and are now around 1e-9). Numdiff does not complain about it on my machine though (the tests pass on my machine).
It looks like some kind of floating-point issue, but I am not sure how this PR could have affected it. 
Is it possible that it was #3066 instead? Because from there on we no longer assign the weights to anything before comparing them, only read them. Still this would be a strange explanation.

@cpraveen: You opened #3013 before this PR was merged, so I would be surprised if it was this PR that introduced the trouble for you. But since @luca-heltai and @nicola-giuliani are already looking into this, I would wait for their opinion. If there are any specific questions you have, I am happy to discuss or look into them.

cpraveen commented:
@luca-heltai @nicola-giuliani I have modified your gist to be closer to my use case, see

https://gist.github.com/cpraveen/6c50ca2594aa0ead9ea2e95dc1e4c6a9

Shall I open a new issue to discuss this ?

kronbichler commented:
@gassmoeller It seems that switching to `std::sort` from #3066 restorted the good properties. So it might have been some subtle issue with floating point comparisons in the linear N^2 search. (I guess my comment was a bit premature...)

gassmoeller commented:
@kronbichler: No problem. Good to know everything is back to normal.
